### PR TITLE
gh-107972: Argument Clinic: Ensure a C basename is provided after 'as'

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-107972

### CODE REVIEW

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -4864,9 +4864,11 @@ def state_modulename_name(self, line: str) -> None:
         before, equals, existing = line.rpartition('=')
         c_basename: str | None
         if equals:
-            full_name, _, c_basename = before.partition(' as ')
+            full_name, _as, c_basename = before.partition(' as ')
```
Comment (AA-Turner) [inline Tools/clinic/clinic.py]: Style, but I've tended to see a leading underscore as 'unused' vs a trailing underscore as 'reusing keyword'. Feel free to reject the suggestion though!

```suggestion
            full_name, as_, c_basename = before.partition(' as ')
```

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -4864,9 +4864,11 @@ def state_modulename_name(self, line: str) -> None:
...
             full_name = full_name.strip()
             c_basename = c_basename.strip()
+            if _as and not c_basename:
```
Comment (AA-Turner) [inline Tools/clinic/clinic.py]:
```suggestion
            if as_ and not c_basename:
```

File: Tools/clinic/clinic.py [hunk 3]
```diff
@@ -4916,10 +4918,11 @@ def state_modulename_name(self, line: str) -> None:
         line, _, returns = line.partition('->')
         returns = returns.strip()
 
-        full_name, _, c_basename = line.partition(' as ')
+        full_name, _as, c_basename = line.partition(' as ')
```
Comment (AA-Turner) [inline Tools/clinic/clinic.py]:
```suggestion
        full_name, as_, c_basename = line.partition(' as ')
```

File: Tools/clinic/clinic.py [hunk 4]
```diff
@@ -4916,10 +4918,11 @@ def state_modulename_name(self, line: str) -> None:
...
         full_name = full_name.strip()
         c_basename = c_basename.strip() or None
-
+        if _as and not c_basename:
```
Comment (AA-Turner) [inline Tools/clinic/clinic.py]:
```suggestion
        if as_ and not c_basename:
```

File: Tools/clinic/clinic.py (hunk 1 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: clinic.py uses both styles :) I have no preference here.

**Review by AlexWaygood — APPROVED** [review]
+1 to Adam's suggestions, but otherwise this LGTM. It appears we _do_ `fail()` currently for at least the second case you're adding a test for, but the error message is pretty nonsensical currently:

```
Couldn't find existing function 'foo2'!
```

**Comment by erlend-aasland** [comment]
> +1 to Adam's suggestions, but otherwise this LGTM. It appears we _do_ `fail()` currently for at least the second case you're adding a test for, but the error message is pretty nonsensical currently:
> 
> ```
> Couldn't find existing function 'foo2'!
> ```

Well, if you'd defined `foo2` first, you would not get that failure; it would silently accept the weird line, but still produce a valid C basename (just using the full name).

**Comment by AlexWaygood** [comment]
> Well, if you'd defined `foo2` first, you would not get that failure; it would silently accept the weird line, but still produce a valid C basename (just using the full name).

Yup, I gathered. Maybe we should do that in the test, actually, so that the _only_ erroneous thing in the test case is the one we're explicitly testing that clinic catches?

**Comment by erlend-aasland** [comment]
Try this weird input; clinic accepts it on `main`:

```c
/*[clinic input]
output everything block
foo2
[clinic start generated code]*/

/*[clinic input]
output everything block
foo as = foo2
[clinic start generated code]*/
```

**Comment by erlend-aasland** [comment]
> Try this weird input; clinic accepts it on `main`:

And with that in mind, we should probably adjust the test case to look exactly like that.

**Comment by erlend-aasland** [comment]
> Yup, I gathered. Maybe we should do that in the test, actually, so that the _only_ erroneous thing in the test case is the one we're explicitly testing that clinic catches?

I did not read this until now; we came to the same conclusion :)

**Comment by erlend-aasland** [comment]
Thanks for the reviews!

**Pull request merged**